LayoutEditor - More tests!

geowar1 opened the pull request.

bobjacobsen commented:
This is in-queue for AppVeyor now.  Once that's done, I'm going to switch it to be against the `master` branch as part of returning to our usual development process.  That'll cause the tests to rerun.

bobjacobsen commented:
AppVeyor against `development` threw errors:
```
classname                                                     name                InnerXml                                                                                                                                           
1616---------                                                     ----                --------                                                                                                                                           
1617jmri.jmrit.operations.setup.OperationsSetupFrameTest          testSetupFrameWrite <error message="Wait AbstractButton with text "HO" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.TimeoutEx...
1618jmri.jmrit.operations.automation.actions.BuildTrainActionTest testActionMessages  <error message="Wait AbstractButton with text "Halt" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.Timeout...
1619jmri.jmrit.timetable.swing.TimeTableFrameTest                 testDriver          <error message="Wait "Item "Station 2" available in combo box." state to be reached (ComponentOperator.WaitStateTimeout)" ty...
```

geowar1 commented:
Ok… **finally** ready to merge… 

Two notes for reviewers: In JemmyUtil there were a bunch of "enterClickAndLeave" methods (for a bunch of different kinds of controls). All of them used the ".doClick" method except the one for JButtons which used the .push operator. So I refactored the one using ".push()" to "enterPushAndLeave" and created a new "enterClickAndLeave" which used ".doClick" like the others. Although this seems like the "right thing to do" I'm more that a little nervous about changing so many non-LayoutEditor sources… Plus test writers expectations that "enterClickAndLeave" is the right method to use when what they should be using is "enterPushAndLeave" instead. Feedback welcome.

The second note is about the note in the confirmJOptionPane method that said:
>
         // the previous version of this message verified the text string
         // if the dialog matched the passed message value.  We need to
         // determine how to do that using Jemmy.

I looked at the source to the JOptionPane and determined that it just splits the messageText (string) at newline characters and creates a JLabel for each line. So I wrote the waitForLabels method to split the messageText at newline characters and create a JLabelOperator to wait for each line. It worked! So I added the waitForLabels call to all the methods that were passed messageText but were ignoring it. My only concern is that the messageText parameter for those methods is no longer being ignored and that it may cause tests that previously passed to fail. Fortunately, this didn't happen… however I am somewhat concerned that this may have caused other unintended issues.

The biggest issue that I had to find and fix was properly saving and restoring the DefaultStringComparator. I had one case (in LevelXingTest.java) that looked like it was doing the right thing… when in actuality it wasn't. It took me forever to track that down but once I did all the non-LayoutEditor tests started passing.

My only remaining concern is how many of the tests that I wrote are timing out. While they run just fine on my machine on the Travis/AppVeyor builds they almost constantly fail. They are currently marked "@Ignore" but I'd like to determine why they're failing on the CI builds.

I have (at least) two questions about Jemmy testing: 
What criteria do I use when choosing between the .push and .doClick methods? Currently I try .push and if that doesn't work I try .doClcik… but I'd like to know when it's appropriate to use one vs. the other. (I see both used in the non-LayoutEditor test sources.)

Also what criteria do I use when choosing between the .clearTest/.typeText, .enterText and .setText methods? Again currently I'm just trying each to see what works. (note: when texting live (as-you type) text input verification I generally use the .clearText/.typeText methods and the .setText method for everything else. But I do see the .enterText methods used occasionally in other test sources. When is it the appropriate method to use?

Your feedback on these issues (and any others you might have) would be appreciated. TIA.

bobjacobsen commented:
Resolved conflicts with master in import section of two files.

bobjacobsen commented:
Close/reopen to pick up fix in #7891

pabender commented:
@geowar1 I'm starting to look through this, but it might take me a day or so to finish.

bobjacobsen commented:
This is now showing non-trivial conflicts, probably after merging #7880
```
java/src/jmri/jmrit/display/layoutEditor/LayoutEditor.java
java/src/jmri/jmrit/display/layoutEditor/LayoutEditorComponent.java
```

bobjacobsen commented:
Manually resolved new conflicts in `java/src/jmri/jmrit/display/layoutEditor/TrackSegment.java`.  Three were just due to whitespace changes.  Remaining one was:
```
<<<<<<< WIP-SUMore-tests
            double distance, minDistance = Double.POSITIVE_INFINITY;
=======
            double distance, minDistance = Float.POSITIVE_INFINITY;
>>>>>>> master
```

Took the version from WIP-SUMore-tests

bobjacobsen commented:
Travis headless restart:
[ERROR] Errors: 
792[ERROR] jmri.jmrix.sprog.SprogCSThrottleTest.testSetF20(jmri.jmrix.sprog.SprogCSThrottleTest)
793[ERROR]   Run 1: SprogCSThrottleTest.setUp:372 » ConcurrentModification
794[ERROR]   Run 2: SprogCSThrottleTest.tearDown:386 NullPointer
795[INFO] 
796[INFO] 
797[ERROR] Tests run: 42367, Failures: 0, Errors: 1, Skipped: 3407

bobjacobsen commented:
Travis graphical:

[ERROR]   LayoutEditorChecksTest.tearDownClass:314 Unexpected ERROR or higher messages emitted: testToolsCheckMenuExists(jmri.jmrit.display.layoutEditor.LayoutEditorChecksTest) : giving up after 4 failures
948[ERROR] Errors: 
949[ERROR]   ConvertDialogTest.testCTorConvert:47 » TimeoutExpired Wait event queue staying...
950[ERROR]   LayoutEditorChecksTest.testToolsCheckMenuExists:69 » TestTimedOut test timed o...
951[ERROR]   LayoutEditorTest.testToolBarPositionFloat:984 » TestTimedOut test timed out af...
952[ERROR]   LayoutEditorTest.testToolBarPositionRight:958 » TestTimedOut test timed out af...
953[ERROR] jmri.jmrix.sprog.SprogCSThrottleTest.testSetF21(jmri.jmrix.sprog.SprogCSThrottleTest)
954[ERROR]   Run 1: SprogCSThrottleTest.setUp:372 » ConcurrentModification
955[ERROR]   Run 2: SprogCSThrottleTest.tearDown:386 NullPointer
956[INFO] 
957[INFO] 
958[ERROR] Tests run: 42380, Failures: 1, Errors: 5, Skipped: 3121
959[INFO]

bobjacobsen commented:
Restarted Travis headless after:
```
[ERROR] Failures: 
796[ERROR] jmri.configurexml.LoadAndStoreTest.loadLoadStoreFileCheck[java/test/jmri/configurexml/load/SectionManagerXmlTest.xml (pass=true)](jmri.configurexml.LoadAndStoreTest)
797[ERROR]   Run 1: LoadAndStoreTest>LoadAndStoreTestBase.loadLoadStoreFileCheck:296->LoadAndStoreTestBase.checkFile:214 expected:<  <[memories class="jmri.managers.configurexml.DefaultMemory]ManagerXml">> but was:<  <[blocks class="jmri.configurexml.Block]ManagerXml">>
798[ERROR]   Run 2: LoadAndStoreTest>LoadAndStoreTestBase.tearDown:317 Unexpected ERROR or higher messages emitted:          to file2:"/home/travis/build/JMRI/JMRI/temp/temp/SectionManagerXmlTest.xml"
```

bobjacobsen commented:
Restarted Travis graphical after:
```
WARN  - Portal IP0001-0002 needs an OBlock on each side [main] jmrit.beantable.oblock.SignalTableModel.makeList()
782Warning: org.jdom2.transform.JDOMSource$DocumentReader: http://javax.xml.XMLConstants/property/accessExternalDTD
783Warning: org.jdom2.transform.JDOMSource$DocumentReader: http://www.oracle.com/xml/jaxp/properties/entityExpansionLimit
784WARN  - testToolBarPositionFloat(jmri.jmrit.display.layoutEditor.LayoutEditorTest) : run  3 failed, RetryRule repeats [main] util.junit.rules.RetryRule.evaluate()
785Disposal was interrupted:
786java.lang.InterruptedException
787	at java.lang.Object.wait(Native Method)
788	at java.lang.Object.wait(Object.java:502)
789	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1343)
790	at java.awt.Window.doDispose(Window.java:1210)
791	at java.awt.Window.dispose(Window.java:1147)
792	at jmri.util.JmriJFrame.dispose(JmriJFrame.java:983)
793	at jmri.jmrit.display.Editor.dispose(Editor.java:2666)
794	at jmri.jmrit.display.IconEditorWindowTest.tearDown(IconEditorWindowTest.java:409)
795	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
796	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
797	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
798	at java.lang.reflect.Method.invoke(Method.java:498)
799	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
800	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
801	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
802	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
803	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
804	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
805	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
806	at java.lang.Thread.run(Thread.java:748)
807Disposal was interrupted:
808java.lang.InterruptedException
809	at java.lang.Object.wait(Native Method)
810	at java.lang.Object.wait(Object.java:502)
811	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1343)
812	at java.awt.Window.doDispose(Window.java:1210)
813	at java.awt.Window.dispose(Window.java:1147)
814	at jmri.util.JmriJFrame.dispose(JmriJFrame.java:983)
815	at jmri.jmrit.display.Editor.dispose(Editor.java:2666)
816	at jmri.jmrit.display.IconEditorWindowTest.tearDown(IconEditorWindowTest.java:409)
817	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
818	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
819	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
820	at java.lang.reflect.Method.invoke(Method.java:498)
821	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
822	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
823	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
824	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
825	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
826	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
827	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
828	at java.lang.Thread.run(Thread.java:748)
829BlueCove log redirected to log4j
830WARN  - notify: During dispatch to jmri.jmrix.sprog.update.SprogIIUpdateFrame[frame485,0,0,242x65,invalid,hidden,layout=java.awt.BorderLayout,title=SPROGII/SPROG 3/Pi-SPROG Firmware Update,resizable,normal,defaultCloseOperation=DISPOSE_ON_CLOSE,rootPane=javax.swing.JRootPane[,0,0,242x65,invalid,layout=javax.swing.JRootPane$RootLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=16777673,maximumSize=,minimumSize=,preferredSize=],rootPaneCheckingEnabled=true] [AWT-EventQueue-1] jmrix.sprog.update.SprogVersionQuery.notifyVersion()
831java.lang.NullPointerException
832	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.requestBoot(SprogIIUpdateFrame.java:275)
833	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.notifyVersion(SprogIIUpdateFrame.java:62)
834	at jmri.jmrix.sprog.update.SprogVersionQuery.notifyVersion(SprogVersionQuery.java:121)
835	at jmri.jmrix.sprog.update.SprogVersionQuery.timeout(SprogVersionQuery.java:271)
836	at jmri.jmrix.sprog.update.SprogVersionQuery$1.actionPerformed(SprogVersionQuery.java:317)
837	at javax.swing.Timer.fireActionPerformed(Timer.java:313)
838	at javax.swing.Timer$DoPostEvent.run(Timer.java:245)
839	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
840	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
841	at java.awt.EventQueue.access$500(EventQueue.java:97)
842	at java.awt.EventQueue$3.run(EventQueue.java:709)
843	at java.awt.EventQueue$3.run(EventQueue.java:703)
844	at java.security.AccessController.doPrivileged(Native Method)
845	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
846	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
847	at org.netbeans.jemmy.QueueTool$JemmyQueue.dispatchEvent(QueueTool.java:656)
848	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
849	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
850	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
851	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
852	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
853	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
854WARN  - notify: During dispatch to jmri.jmrix.sprog.update.SprogIIUpdateFrame[frame485,0,0,242x65,invalid,hidden,layout=java.awt.BorderLayout,title=SPROGII/SPROG 3/Pi-SPROG Firmware Update,resizable,normal,defaultCloseOperation=DISPOSE_ON_CLOSE,rootPane=javax.swing.JRootPane[,0,0,242x65,invalid,layout=javax.swing.JRootPane$RootLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=16777673,maximumSize=,minimumSize=,preferredSize=],rootPaneCheckingEnabled=true] [AWT-EventQueue-1] jmrix.sprog.update.SprogVersionQuery.notifyVersion()
855java.lang.NullPointerException
856	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.requestBoot(SprogIIUpdateFrame.java:275)
857	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.notifyVersion(SprogIIUpdateFrame.java:62)
858	at jmri.jmrix.sprog.update.SprogVersionQuery.notifyVersion(SprogVersionQuery.java:121)
859	at jmri.jmrix.sprog.update.SprogVersionQuery.timeout(SprogVersionQuery.java:271)
860	at jmri.jmrix.sprog.update.SprogVersionQuery$1.actionPerformed(SprogVersionQuery.java:317)
861	at javax.swing.Timer.fireActionPerformed(Timer.java:313)
862	at javax.swing.Timer$DoPostEvent.run(Timer.java:245)
863	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
864	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
865	at java.awt.EventQueue.access$500(EventQueue.java:97)
866	at java.awt.EventQueue$3.run(EventQueue.java:709)
867	at java.awt.EventQueue$3.run(EventQueue.java:703)
868	at java.security.AccessController.doPrivileged(Native Method)
869	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
870	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
871	at org.netbeans.jemmy.QueueTool$JemmyQueue.dispatchEvent(QueueTool.java:656)
872	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
873	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
874	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
875	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
876	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
877	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
878WARN  - notify: During dispatch to jmri.jmrix.sprog.update.SprogIIUpdateFrame[frame485,0,0,242x65,invalid,hidden,layout=java.awt.BorderLayout,title=SPROGII/SPROG 3/Pi-SPROG Firmware Update,resizable,normal,defaultCloseOperation=DISPOSE_ON_CLOSE,rootPane=javax.swing.JRootPane[,0,0,242x65,invalid,layout=javax.swing.JRootPane$RootLayout,alignmentX=0.0,alignmentY=0.0,border=,flags=16777673,maximumSize=,minimumSize=,preferredSize=],rootPaneCheckingEnabled=true] [AWT-EventQueue-1] jmrix.sprog.update.SprogVersionQuery.notifyVersion()
879java.lang.NullPointerException
880	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.requestBoot(SprogIIUpdateFrame.java:275)
881	at jmri.jmrix.sprog.update.SprogIIUpdateFrame.notifyVersion(SprogIIUpdateFrame.java:62)
882	at jmri.jmrix.sprog.update.SprogVersionQuery.notifyVersion(SprogVersionQuery.java:121)
883	at jmri.jmrix.sprog.update.SprogVersionQuery.timeout(SprogVersionQuery.java:271)
884	at jmri.jmrix.sprog.update.SprogVersionQuery$1.actionPerformed(SprogVersionQuery.java:317)
885	at javax.swing.Timer.fireActionPerformed(Timer.java:313)
886	at javax.swing.Timer$DoPostEvent.run(Timer.java:245)
887	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
888	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
889	at java.awt.EventQueue.access$500(EventQueue.java:97)
890	at java.awt.EventQueue$3.run(EventQueue.java:709)
891	at java.awt.EventQueue$3.run(EventQueue.java:703)
892	at java.security.AccessController.doPrivileged(Native Method)
893	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
894	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
895	at org.netbeans.jemmy.QueueTool$JemmyQueue.dispatchEvent(QueueTool.java:656)
896	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
897	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
898	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
899	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
900	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
901	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
902INFO  - Tests running on JMRI 4.19.3ish+travis+2020-01-15T13:31:31Z+R186be39 with Java 1.8.0_222 from en_US [main] jmri.VersionTest.announceVersions()
903[ERROR] Failures: 
904[ERROR]   LevelXingTest.testGetCoordsForConnectionType:334 Looking for ERROR message "Invalid connection type 0" got "getCoordsForConnectionType(): Invalid connection type 0"
905[ERROR]   LevelXingTest.testSetConnections:267 Looking for message "Invalid Connection Type 0" and didn't find it
906[INFO] 
907[ERROR] Tests run: 42401, Failures: 2, Errors: 0, Skipped: 3129
908
```

bobjacobsen commented:
Restarted AppVeyor after:
```
jmri.jmrit.display.layoutEditor.LevelXingTest testGetCoordsForConnectionType <failure message="Looking for ERROR message "Invalid connection type 0" got "getCoordsForConnectionType(): Invalid connection type 0&q...
3404jmri.jmrit.display.layoutEditor.LevelXingTest testSetConnections             <failure message="Looking for message "Invalid Connection Type 0" and didn't find it" type="junit.framework.AssertionFailedError">junit.fra...
```

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorToolsTest.java:
```diff
@@ -487,92 +1116,128 @@ public void testFinalizeBlockBossLogicNullInput() {
 
     @Test
-    @Ignore("causes error on jenkins; exhausts failure retries")
+    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
```
There isn't any change here other than commenting out the ignore annotation.  What changed to make this work reliably?

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -731,12 +730,12 @@ public Rectangle2D getBounds() {
      * Add Layout Blocks.
      */
-    @SuppressFBWarnings(value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE", justification = "Null is accepted as a valid value")
+    ///@SuppressFBWarnings(value = "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE", justification = "Null is accepted as a valid value")
```
I'm a firm believer in NOT suppressing warnings/errors… I disabled this to determine what the actual error was that's being suppressed so I could fix it… or re-enable the warning if I couldn't determine how to fix it. I'm leaving it here as long as this is a WIP. If/when this ever passes CI I can go back and remove or re-enable any/all of these that weren't resolved.

geowar1 commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorTest.java:
```diff
@@ -921,7 +920,7 @@ public void testToolBarPositionLeft() {
 
     @Test
-    @Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
+    ///@Ignore("Fails on Travis 12/20/2019")
```
WIP: commented out to determine if it's still failing. NOTE: I use the triple-slash for things to remove/re-enable before production/merging. WIP will not be removed until these are all resolved.

geowar1 replied:
find java/test -type f -name '*Test.java' -exec grep -li --null '///@' {} \; | xargs -0t -n 1 grep -i --color=auto '///@'
grep -i --color=auto ///@ java/test/jmri/jmrit/display/layoutEditor/LayoutEditorDialogs/LayoutTrackEditorsTest.java
    ///@Ignore("giving up after X failures")
grep -i --color=auto ///@ java/test/jmri/jmrit/display/layoutEditor/LayoutEditorTest.java
    ///@Ignore("unreliable on CI servers")
    ///@Ignore("Fails on Travis 12/20/2019")
    ///@Ignore("Fails on Travis 12/20/2019")
    ///@Ignore("Fails on Travis 12/20/2019")
    ///@Ignore("Fails on Travis 12/20/2019")
grep -i --color=auto ///@ java/test/jmri/jmrit/display/layoutEditor/LayoutEditorToolsTest.java
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")
    ///@Ignore("Fails on AppVeyor, macOS and Windows 12/20/2019")

These will all be ether re-enabled or removed before WIP is removed.

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -633,28 +632,28 @@ public LayoutTrack getConnectD() {
     public void setConnectA(LayoutTrack o, int type) {
         connectA = o;
-        if ((connectA != null) && (type != TRACK)) {
+        if ((connectA == null) ? (type != NONE) : ((!(o instanceof TrackSegment)) || (type != TRACK))) {
```
https://guide.freecodecamp.org/c/ternary-operator/

geowar1 replied:
`      if (connectA == null) {
            if (type != NONE) {
                log.error("unexpected type of A connection to levelXing - " + type);
            }
        } else {
            if ((!(o instanceof TrackSegment)) || (type != TRACK)) {
                log.error("unexpected type of A connection to levelXing - " + type);
            }
        }
`
So 10 lines of code instead of 3… and DRY fail.

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -633,28 +632,28 @@ public LayoutTrack getConnectD() {
     public void setConnectD(LayoutTrack o, int type) {
         connectD = o;
-        if ((connectD != null) && (type != TRACK)) {
+        if ((connectD == null) ? (type != NONE) : ((!(o instanceof TrackSegment)) || (type != TRACK))) {
```
Is this better?
```
        if (((connectD == null) && (type != NONE))
                || ((connectD != null) && (o instanceof TrackSegment) && (type != TRACK))) {
            log.error("unexpected type of D connection to levelXing - " + type);
        }

```
Want to guess which one code climate complains about as being too complex?

geowar1 commented:
@pabender What I really need is for you to tell me why these tests are failing and how they can be fixed.

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -633,28 +632,28 @@ public LayoutTrack getConnectD() {
     public void setConnectB(LayoutTrack o, int type) {
         connectB = o;
-        if ((connectB != null) && (type != TRACK)) {
+        if ((connectB == null) ? (type != NONE) : ((!(o instanceof TrackSegment)) || (type != TRACK))) {
```
All ternary operators removed

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -633,28 +632,28 @@ public LayoutTrack getConnectD() {
     public void setConnectB(LayoutTrack o, int type) {
         connectB = o;
-        if ((connectB != null) && (type != TRACK)) {
+        if ((connectB == null) ? (type != NONE) : ((!(o instanceof TrackSegment)) || (type != TRACK))) {
```
Ternary operator removed

geowar1 commented on java/src/jmri/jmrit/display/layoutEditor/LevelXing.java:
```diff
@@ -633,28 +632,28 @@ public LayoutTrack getConnectD() {
     public void setConnectC(LayoutTrack o, int type) {
         connectC = o;
-        if ((connectC != null) && (type != TRACK)) {
+        if ((connectC == null) ? (type != NONE) : ((!(o instanceof TrackSegment)) || (type != TRACK))) {
```
Ternary operator removed

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorChecksTest.java:
```diff
@@ -1,40 +1,316 @@
+    }   //testCheckUnnecessaryAnchors
+
+    @BeforeClass
+    public static void setUpClass() {
```
@BeforeClass and @AfterClass are supposed to be used for tasks that take a significant amount of time, such as setting up a database.

Starting layout editor does NOT take a significant amount of time, even if you configure items in the layout editor.

When you only configure the layout editor once for all tests in the file, then any operation you perform in one test changes the initial state for the next test, which means the tests Katy be order dependent.

This could very well impact why some tests fail in some environment and not others.

pabender commented:
> @pabender What I really need is for you to tell me why these tests are failing and how they can be fixed.

Until I can figure out why alltest is hanging on Jenkins, I don’t have the bandwidth to investigate these tests too.

pabender commented:
@geowar1 when you get a chance, can you pull master into this branch so we get some of the more recent changes to the testing infrastructure

geowar1 commented:
Why can't I resave the conflicts on this PR?
